packages/yew-router/src/hidden.rs

```diff
@@ -0,0 +1,407 @@
+        // candidate routes.
+        let regex_set = cached_regex_set::<T>();
+        for candidate in regex_set.matches(&route.path) {
+            // Try each route in turn. `RegexSet` guarantees iteration order to
+            // match the original array.
+            let variant = &T::VARIANTS[candidate];
+            if let Some(res) = variant.from_route(route) {
+                return Ok(res);
+            }
+        }
+        Err(T::default())
+    }
+
+    fn to_route(&self) -> Route {
+        let (index, args) = self.to_args();
+        let variant = &T::VARIANTS[index];
+        variant.reconstitute(args)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Arg {
+    Empty,
+    String(String),
+    Opaque(JsValue),
+}
+
+impl From<String> for Arg {
+    fn from(s: String) -> Self {
+        Arg::String(s)
+    }
+}
+
+impl From<&str> for Arg {
+    fn from(s: &str) -> Self {
+        Arg::String(s.into())
+    }
+}
+
+impl From<JsValue> for Arg {
+    fn from(v: JsValue) -> Self {
+        if v.is_undefined() {
+            Arg::Empty
+        } else {
+            Arg::Opaque(v)
+        }
+    }
+}
+
+impl TryFrom<Arg> for String {
+    type Error = ();
+
+    fn try_from(value: Arg) -> Result<Self, Self::Error> {
+        if let Arg::String(s) = value {
+            Ok(s)
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl TryFrom<Arg> for JsValue {
+    type Error = ();
+
+    fn try_from(value: Arg) -> Result<Self, Self::Error> {
+        if let Arg::Opaque(v) = value {
+            Ok(v)
+        } else {
+            Err(())
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Args {
+    args: Vec<Arg>,
+    route: Route,
+}
+
+impl Args {
+    pub fn new(route: Route) -> Self {
+        Self {
+            route,
+            args: Vec::new(),
+        }
+    }
+    pub fn empty() -> Self {
+        Self::new(Route::new("/"))
+    }
+    pub fn push<T: Into<Arg>>(&mut self, arg: T) {
+        self.args.push(arg.into());
+    }
+    pub fn pop<T>(&mut self) -> Option<T>
+    where
+        Arg: TryInto<T>,
+    {
+        self.args.pop().expect("No argument to pop").try_into().ok()
+    }
+    pub fn push_serde<T: Serialize>(&mut self, arg: &T) {
+        self.push(JsValue::from_serde(arg).expect("Argument to serialize"));
+    }
+    pub fn pop_serde<T: DeserializeOwned>(&mut self) -> Option<T> {
+        self.pop::<JsValue>()?.into_serde().ok()
+    }
+    pub fn push_str<T: Display>(&mut self, arg: &T) {
+        self.push(arg.to_string());
+    }
+    pub fn pop_str<T: FromStr>(&mut self) -> Option<T> {
+        self.pop::<String>()?.parse().ok()
+    }
+    pub fn take_route(&mut self) -> Route {
+        mem::replace(&mut self.route, Default::default())
+    }
+    pub fn store_route(&mut self, route: Route) {
+        self.route = route;
+    }
+}
+
+/// An instance of this type is constructed for each variant by the
+/// derive macro.
+pub struct RouteMuncher<T: DerivedRoutable> {
```

Comment:
Do we have a name better than "Muncher"? I can't even fully understand what "Muncher" means here.

Reply:
Happy to rename it if you have a better name. It eats part of the `Route` leaving the rest untouched, hence "muncher".

Reply:
I can't think of any right now but I would like to have it renamed before merging. "muncher" seems confusing to me

Reply:
Some ideas:

- `ArgumentPlucker`
- `RouteProcessor`
- `BindingGroup`

Reply:
I'd say either `RouteProcessor` or `RouteParser`. One question though: would it be possible to do the parsing for routes in the macro? That way, we can reduce the computation done in browser, on the UI thread.

Reply:
You mean parsing the `#[at(..)]` path? That's already done by the macro.

Reply:
Actually, that function is only called on the current URL, it's not used on the "at" path.

Reply:
Oh, I guess it stays how it currently is then

Reply:
I've done the renaming.